code/ATMOSPHERICS/hvac/geyser.dm

```diff
@@ -0,0 +1,46 @@
+#define GEYSER_TEMP 40
+#define GEYSER_POWER 40000
+#define GEYSER_HEAT_RANGE 3
+#define GEYSER_WARM_RATE 25
+
+/obj/structure/geyser
+	name = "thermal geyser"
+	desc = "A geyser is a type of spring that is caused by the upward force of heat attemping to escape outward from a mineral body."
+	density = 1
+	anchored = 1
+	icon = 'icons/misc/beach.dmi'
+	icon_state = "geyser"
+
+/obj/structure/geyser/process()
+	if(prob(40))
+		var/datum/effect/effect/system/smoke_spread/smoke = new /datum/effect/effect/system/smoke_spread()
+		smoke.set_up(2, 0, get_turf(src)) //Make 2 drifting clouds of smoke, direction
+		smoke.time_to_live = 2 SECONDS //unusually short smoke
+		smoke.start()
+	var/turf/T = loc
+	if(!istype(T))
+		return //Not on a turf
+	//If we're on a simulated turf, just heat the air.
+	if(istype(T,/turf/simulated))
```

Comment:
Rather than all this stuff, just check if it's on a turf, if so get the gas_mixture like you have done so, and call gas_mixture.add_thermal_energy(however much energy you want to add in joules)

Unsimulated add_thermal_energy just returns 0, so you won't be heating up space with these.

Reply:
Part of what this code (copied from space heaters, I think?) does is make sure the temperature doesn't exceed the cap
The proper way to do this is something like `add_thermal_energy(min(energy_to_add, get_thermal_energy_change(max_temp)))`
This is extremely ugly and involves a lot of unnecessary calculations so at some point I'll probably add either a capped version of `add_thermal_energy()` or just add a cap parameter to it